### PR TITLE
Better Supermatter (real)

### PR DESCRIPTION
**Pull request by whateverusername0**
## About the PR

Massively refactored SupermatterSystem because in my opinion it looked knda stinky. Made it look more readable and eye appealing :3
- Supermatter now shoots lightning bolts depending on the amount of power it currently has.
- Is able to delaminate into a tesla.
- Starts dormant round-start. To activate it, you need to throw anything at it, or, basically, make it collide with something.
- Makes global announcements whether it's delamming or not, changes station code to whatever if delaminating, which is ofc reversible if there are crazy engineers who will prevent, or at least contain the delam.
- And it now shows integrity when examined within an appropriate range.

Added new objective - cut off a sliver from the supermatter crystal. Basically you have to use a knife or whatever's sharp on the SM crystal, (which will activate it, if it was dormant), stand still for 30 seconds, and run away with a highly radioactive piece of the supermatter. This will greatly damage the SM and alert the crew with a massive fucking announcement stating that "SOMEONE HAS TAMPERED WITH SM ENGINEERS GET HIS ASS!!"

Added some placeholders for cascade resonance delamination type if nobilium ever gets added or smh else comes up on my mind.

And mapped the Supermatter chamber on Fland, removing the Singuloo platform.

## Why / Balance

because i fucking love supermatter and because i can

## Media

fland now having SM
image

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## TODO
- [X] Playtest and bug fix.

**Changelog**

:cl:
- add: Supermatter now shoots out lightning bolts.
- add: You will now have to activate the SM by making it collide with anything.
- add: Supermatter Tesla Delamination.
- add: New traitor objective - get a piece of the supermatter.
- remove: Fland: Singularity platform.
- add: Fland: Supermatter chamber.
- tweak: Fland: Moved TEG near the SM chambere.

### CODE REVIEW

File: Resources/Prototypes/Goobstation/Entities/Objects/Misc/supermatter_sliver.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  name: supermatter sliver
+  description: A shard from the station's supermatter crystal. Highly radioactive.
+  components:
+    - type: PointLight
+      enabled: true
+      radius: 3
+      energy: 2
+      color: "#fff633"
+    - type: RadiationSource
+      intensity: 3.5
```
Comment (misandrie) [inline Resources/Prototypes/Goobstation/Entities/Objects/Misc/supermatter_sliver.yml:13]: While sure, makes sense - this requires a container to mask both of these components - else you're sentencing a would-be thief to be in a rad suit the moment he picks up the sliver

File: Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs [hunk 2]
```diff
@@ -73,444 +92,526 @@ public override void Update(float frameTime)
...
+        /// <param name="customSender">If true, sends the announcement from Central Command.</param>
+        public void SupermatterAnnouncement(EntityUid uid, string message, bool global = false, bool customSender = false)
+        {
+            if (global)
+            {
+                _chat.DispatchStationAnnouncement(uid, message, customSender ? "Central Command" : Loc.GetString("supermatter-announcer"), colorOverride: Color.Yellow);
+                return;
+            }
+            _chat.TrySendInGameICMessage(uid, message, InGameICChatType.Speak, hideChat: false, checkRadioPrefix: true);
         }
```
Comment (misandrie) [inline Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs:419]: If you're doing "customSender" - don't make it "Central Command" if true, that's not really *custom*.
make customSender a nullable string and use null-coalescing.

File: Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs [hunk 3]
```diff
@@ -73,444 +92,526 @@ public override void Update(float frameTime)
...
+        }
+
+        private void OnGetSliver(EntityUid uid, SupermatterComponent sm, ref SupermatterDoAfterEvent args)
+        {
+            sm.Damage += sm.DelaminationPoint / 10; // your criminal actions will not go unnoticed
+            SupermatterAnnouncement(uid, Loc.GetString("supermatter-announcement-cc-tamper", ("integrity", GetIntegrity(sm.Damage, sm.DelaminationPoint).ToString("0.00"))), true, true);
+
+            Spawn(sm.SliverPrototypeId, _transform.GetMapCoordinates(args.User));
+            _popup.PopupClient(Loc.GetString("supermatter-tamper-end"), args.User);
+        }
```
Comment (misandrie) [inline Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs:604]: Delay this, SM/Engi is getting swarmed immediately as this is announced.

File: Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs [hunk 4]
```diff
@@ -73,444 +92,526 @@ public override void Update(float frameTime)
...
-            var integrity = damage / explosionPoint;
+            var integrity = damage / maxDamage;
             integrity = (float) Math.Round(100 - integrity * 100, 2);
             integrity = integrity < 0 ? 0 : integrity;
             return integrity;
         }
 
         /// <summary>
-        /// Runs the logic and timers for Delamination
+        ///     The supermatter is picky. Help the SM pick it's last song of death.
```
Comment (misandrie) [inline Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs:433]: Ambigious comment

**Review by misandrie — CHANGES_REQUESTED** [review]
Offer these changes to the [upstream where the code is ported from](https://github.com/Simple-Station/Einstein-Engines/pull/341) for better scrutiny

Not really going into the unneccesary variable name changes (they were explicit for a reason).

**Comment by VMSolidus** [comment]
Hey uhh, as this code's maintainer and current codeowner, I'd appreciate it if you brought your changes to the Einstein Engines Upstream where this system is being maintained(and is licensed from). It would be for the betterment of all servers that use our Supermatter Engines if improvements to the code were made more wide reaching. I'm actually completely willing to work with you directly on making these changes directly to the base code, and ensuring that this code is in the best state it can be for its eventual release to the game. Here's the link to where I have the code currently. I think most of your changes are actually great, especially in adding the traitor objective, optional tesla delamination, and improving the delamination logic in general, but I also have even further improvements planned that I'd love to see implemented. For example I want to have hotloaded CVars affecting the delamination logic, so that a server host can optionally override the delamination type to force a specific kind of event to happen, or a boolean that bypasses the checking logic and forces a "Random-Any" delamination regardless of the conditions leading up to it. 

https://github.com/Simple-Station/Einstein-Engines/pull/341

File: Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs (hunk 3 above)
Comment (whateverusername0) [reply Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs:604]: > Delay this, SM/Engi is getting swarmed immediately as this is announced.

i honestly dunno how

**Comment by whateverusername0** [comment]
> Hey uhh, as this code's maintainer and current codeowner, I'd appreciate it if you brought your changes to the Einstein Engines Upstream where this system is being maintained(and is licensed from).

sure. do i just make a PR on your branch?

**Review by Aidenkrz — CHANGES_REQUESTED** [review]
Mary has already stated most of the issues with this, fix em up and fix the checks, otherwise this looks good.

File: Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs (hunk 4 above)
Comment (whateverusername0) [reply Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs]: lrp comment now

File: Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs (hunk 3 above)
Comment (whateverusername0) [reply Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs]: upped the difficulty to be on the nuke disk level

File: Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs (hunk 2 above)
Comment (whateverusername0) [reply Content.Server/Goobstation/Supermatter/Systems/SupermatterSystem.cs]: done

**Review by Aidenkrz — APPROVED** [review]

**Comment by CatMagic1** [comment]
What the fuck when did we start getting real coders

**Comment by misandrie** [comment]
Don't worry about it

**Pull request merged**